File: webapp/js/openingHours.js
// Sehr einfache Auswertung des OSM `opening_hours`-Tags.
// Deckt gängige Muster ab (z.B. "Mo-Fr 10:00-22:00; Sa-Su 12:00-23:00").
// Komplexere Syntax (PH, Feiertage, Kommentare) wird als "unknown" behandelt.

const DAY_NAMES = ['Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su'];

function dayMatches(dayPart, today) {
  for (const segment of dayPart.split(',')) {
    if (segment.includes('-')) {
      const [from, to] = segment.split('-');
      const startIdx = DAY_NAMES.indexOf(from);
      const endIdx = DAY_NAMES.indexOf(to);
      const todayIdx = DAY_NAMES.indexOf(today);
      if (startIdx === -1 || endIdx === -1 || todayIdx === -1) continue;
      if (startIdx <= endIdx) {
        if (todayIdx >= startIdx && todayIdx <= endIdx) return true;
      } else {
        if (todayIdx >= startIdx || todayIdx <= endIdx) return true;
      }
    } else if (segment === today) {
      return true;
    }
  }
  return false;
}

function parseTime(value) {
  const match = /^(\d{1,2}):(\d{2})$/.exec(value.trim());
  if (!match) return null;
  return parseInt(match[1], 10) * 60 + parseInt(match[2], 10);
}

// returns 'open' | 'closed' | 'unknown'
function currentOpenStatus(openingHours, now) {
  if (!openingHours || !openingHours.trim()) return 'unknown';
  const raw = openingHours.trim();
  if (raw === '24/7') return 'open';

  try {
    const today = DAY_NAMES[now.getDay() === 0 ? 6 : now.getDay() - 1];
    const rules = raw.split(';').map(r => r.trim()).filter(Boolean);
    let matchedAnyRuleForToday = false;

    for (const rule of rules) {
      const parts = rule.split(/\s+/);
      if (parts.length < 2) continue;
      const dayPart = parts[0];
      const timePart = parts.slice(1).join(' ');
      if (!dayMatches(dayPart, today)) continue;
      matchedAnyRuleForToday = true;
      if (/^(off|closed)$/i.test(timePart)) continue;

      for (const span of timePart.split(',')) {
        const times = span.split('-');
        if (times.length !== 2) continue;
        const start = parseTime(times[0]);
        const end = parseTime(times[1]);
        if (start === null || end === null) continue;
        const nowMinutes = now.getHours() * 60 + now.getMinutes();
        if (end > start) {
          if (nowMinutes >= start && nowMinutes < end) return 'open';
        } else {
          if (nowMinutes >= start || nowMinutes < end) return 'open';
        }
      }
    }
    return matchedAnyRuleForToday ? 'closed' : 'unknown';
  } catch (e) {
    return 'unknown';
  }
}

File: bierkarte_app/lib/services/opening_hours.dart
/// Sehr einfache Auswertung des OSM `opening_hours`-Tags.
/// Deckt die häufigsten Muster ab (z.B. "Mo-Fr 10:00-22:00; Sa-Su 12:00-23:00").
/// Komplexere Syntax (PH, Feiertage, Kommentare) wird als "unbekannt" behandelt.
enum OpenStatus { open, closed, unknown }

const _dayNames = ['Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su'];

OpenStatus currentOpenStatus(String? openingHours, DateTime now) {
  if (openingHours == null || openingHours.trim().isEmpty) {
    return OpenStatus.unknown;
  }
  final raw = openingHours.trim();
  if (raw == '24/7') return OpenStatus.open;

  try {
    final today = _dayNames[now.weekday - 1];
    final rules = raw.split(';').map((r) => r.trim()).where((r) => r.isNotEmpty);
    bool matchedAnyRuleForToday = false;
    for (final rule in rules) {
      final parts = rule.split(RegExp(r'\s+'));
      if (parts.length < 2) continue;
      final dayPart = parts[0];
      final timePart = parts.sublist(1).join(' ');
      if (!_dayMatches(dayPart, today)) continue;
      matchedAnyRuleForToday = true;
      if (timePart.toLowerCase() == 'off' || timePart.toLowerCase() == 'closed') {
        continue;
      }
      for (final span in timePart.split(',')) {
        final times = span.split('-');
        if (times.length != 2) continue;
        final start = _parseTime(times[0]);
        final end = _parseTime(times[1]);
        if (start == null || end == null) continue;
        final nowMinutes = now.hour * 60 + now.minute;
        if (end > start) {
          if (nowMinutes >= start && nowMinutes < end) return OpenStatus.open;
        } else {
          // Über Mitternacht hinaus geöffnet (z.B. 18:00-02:00)
          if (nowMinutes >= start || nowMinutes < end) return OpenStatus.open;
        }
      }
    }
    return matchedAnyRuleForToday ? OpenStatus.closed : OpenStatus.unknown;
  } catch (_) {
    return OpenStatus.unknown;
  }
}

bool _dayMatches(String dayPart, String today) {
  for (final segment in dayPart.split(',')) {
    if (segment.contains('-')) {
      final range = segment.split('-');
      if (range.length != 2) continue;
      final startIdx = _dayNames.indexOf(range[0]);
      final endIdx = _dayNames.indexOf(range[1]);
      final todayIdx = _dayNames.indexOf(today);
      if (startIdx == -1 || endIdx == -1 || todayIdx == -1) continue;
      if (startIdx <= endIdx) {
        if (todayIdx >= startIdx && todayIdx <= endIdx) return true;
      } else {
        if (todayIdx >= startIdx || todayIdx <= endIdx) return true;
      }
    } else if (segment == today) {
      return true;
    }
  }
  return false;
}

int? _parseTime(String value) {
  final match = RegExp(r'^(\d{1,2}):(\d{2})$').firstMatch(value.trim());
  if (match == null) return null;
  final hours = int.parse(match.group(1)!);
  final minutes = int.parse(match.group(2)!);
  return hours * 60 + minutes;
}
